### PR TITLE
[#12048] Migrate get feedback responses action

### PR DESCRIPTION
**Pull request by cedricongjh**
Part of #12048 

**Outline of Solution**
- Created factory method for `FeedbackResponse`, similar to `FeedbackQuestion`
- Changed `receiver` to `recipient` and `receiverSection` to `recipientSection` in `FeedbackResponse` and `FeedbackResponseComment`
- Added `FeedbackResponse` and `FeedbackResponseComments` to data bundle
- Added `getFeedbackResponsesFromGiverForQuestion` in `FeedbackResponsesDb`
- Added  `getFeedbackResponseCommentForResponseFromParticipant` in `FeedbackresponseCommentsDb`

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/GetFeedbackResponsesAction.java [hunk 1]
```diff
@@ -62,15 +99,62 @@ public JsonResult execute() {
...
+                    responsesData.add(data);
+                    return;
+                }
+                // Only MCQ and MSQ questions can have participant comment
+                FeedbackResponseCommentAttributes comment =
+                        logic.getFeedbackResponseCommentForResponseFromParticipant(response.getId());
+                if (comment != null) {
+                    data.setGiverComment(new FeedbackResponseCommentData(comment));
+                }
+                responsesData.add(data);
```
Comment (hhdqirui) [inline src/main/java/teammates/ui/webapi/GetFeedbackResponsesAction.java:132]: This is correct but may be unnecessary. We will leave it for now.

File: src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java [hunk 2]
```diff
@@ -229,7 +229,7 @@ protected void testAccessControl_getNonExistingFeedbackResponse_shouldFail() {
                 Const.ParamsNames.FEEDBACK_QUESTION_ID, "randomNonExistId",
                 Const.ParamsNames.INTENT, Intent.INSTRUCTOR_SUBMISSION.toString(),
         };
-        verifyEntityNotFoundAcl(nonExistParams);
+        verifyHttpParameterFailureAcl(nonExistParams);
```
Comment (hhdqirui) [inline src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java:232]: Any idea why it gives a different error?

File: src/main/java/teammates/storage/sqlentity/FeedbackResponse.java [hunk 3]
```diff
@@ -65,18 +68,83 @@ protected FeedbackResponse() {
...
+    /**
+     * Creates a feedback response according to its {@code FeedbackQuestionType}.
+     */
+    public static FeedbackResponse makeResponse(
+            FeedbackQuestion feedbackQuestion, String giver,
+            Section giverSection, String receiver, Section receiverSection,
+            FeedbackResponseDetails responseDetails
+    ) {
+        FeedbackResponse feedbackResponse = null;
+        switch (responseDetails.getQuestionType()) {
```
Comment (hhdqirui) [inline src/main/java/teammates/storage/sqlentity/FeedbackResponse.java:91]: Please also change to `recipient` for this method.

File: src/main/java/teammates/storage/sqlentity/responses/FeedbackConstantSumResponse.java [hunk 4]
```diff
@@ -22,6 +25,15 @@ protected FeedbackConstantSumResponse() {
...
 
+    public FeedbackConstantSumResponse(
+            FeedbackQuestion feedbackQuestion, String giver,
+            Section giverSection, String receiver, Section receiverSection,
+            FeedbackResponseDetails responseDetails
+    ) {
+        super(feedbackQuestion, giver, giverSection, receiver, receiverSection);
+        this.setAnswer((FeedbackConstantSumResponseDetails) responseDetails);
+    }
+
```
Comment (hhdqirui) [inline src/main/java/teammates/storage/sqlentity/responses/FeedbackConstantSumResponse.java:36]: Please also this to `recipient`. Same for all the other feedback response classes.

**Review by hhdqirui — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java (hunk 2 above)
Comment (cedricongjh) [reply src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java:232]: For this test, it was previously for `entityNotFound` for a questionId. 

We call the old logic to try to fetch the question, which is null. 
Then, we try to call the new logic to fetch the question, but there will be a http param error instead as the old questionId is `String`, while the new questionId is `UUID`

File: src/main/java/teammates/storage/sqlentity/FeedbackResponse.java (hunk 3 above)
Comment (cedricongjh) [reply src/main/java/teammates/storage/sqlentity/FeedbackResponse.java:91]: changed!

File: src/main/java/teammates/storage/sqlentity/responses/FeedbackConstantSumResponse.java (hunk 4 above)
Comment (cedricongjh) [reply src/main/java/teammates/storage/sqlentity/responses/FeedbackConstantSumResponse.java]: changed!

**Review by hhdqirui — APPROVED** [review]

File: src/main/java/teammates/ui/webapi/GetFeedbackResponsesAction.java [hunk 5]
```diff
@@ -62,15 +99,62 @@ public JsonResult execute() {
         Intent intent = Intent.valueOf(getNonNullRequestParamValue(Const.ParamsNames.INTENT));
         FeedbackQuestionAttributes questionAttributes = logic.getFeedbackQuestion(feedbackQuestionId);
 
-        List<FeedbackResponseAttributes> responses;
+        if (questionAttributes != null) {
```
Comment (samuelfangjw) [inline src/main/java/teammates/ui/webapi/GetFeedbackResponsesAction.java:102]: Whether or not an entity is migrated is solely dependent on the isMigrated flag in Course.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java (hunk 2 above)
Comment (samuelfangjw) [reply src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java:232]: There should be no need for any changes to behaviour of existing API for GET actions.

**Review by hhdqirui — APPROVED** [review]

**Pull request merged**